base/dict.jl

```diff
@@ -392,6 +392,47 @@ function setindex!(h::Dict{K,V}, v0, key::K) where V where K
     return h
 end
 
+function modify!(f, h::Dict{K, V}, key0) where {K, V}
+    key = convert(K, key0)
+    if !isequal(key, key0)
+        throw(ArgumentError("$(limitrepr(key0)) is not a valid key for type $K"))
+    end
+
+    # Ideally, to improve performance for the case that requires
+    # resizing, we should use something like `ht_keyindex` while
+    # keeping computed hash value and then do something like
+    # `ht_keyindex2!` if `f` returns non-`nothing`.
+    idx = ht_keyindex2!(h, key)
+
+    age0 = h.age
+    if idx > 0
+        @inbounds vold = h.vals[idx]
+        vold = Some{V}(vold)
+        vnew = f(vold)
+    else
+        vold = nothing
+        vnew = f(vold)
+    end
+    if h.age != age0
+        idx = ht_keyindex2!(h, key)
+    end
+
+    if vnew === nothing
+        if idx > 0
+            _delete!(h, idx)
+        end
+    else
+        if idx > 0
+            h.age += 1
+            @inbounds h.keys[idx] = key
+            @inbounds h.vals[idx] = something(vnew)
+        else
+            @inbounds _setindex!(h, something(vnew), key, -idx)
+        end
+    end
+    return vold => vnew
```

Comment:
While playing around with the code, I noticed that this line allocates unless I define `modify!` with `@inline`.

I also noticed that the code runs a bit slower on master compared to 1.8.2 and 1.9.0-beta2 (9ns vs 8ns, always using `@inline`).
<details><summary>Code used</summary>

```
# add m to h[key] without storing zero values
function add!(h, key, m)
    modify!(h, key) do sn
        n = something(sn, 0) + m
        iszero(n) ? nothing : n
    end
    h
end

t = Dict('a' => 50, 'b' => 2)

@btime add!(tt, 'a', 1) setup = (tt = copy($t))
```
</details>